src/open_hash_set.h

```diff
@@ -0,0 +1,106 @@
+#ifndef OPEN_HASH_MAP_H
+#define OPEN_HASH_MAP_H
+
+#include <utility>
+
+#include <stdio.h>
+
+/** Implements an open hash set.
+ */
+template<class Key, class IsKeyNull, class Hash = std::hash<Key>, class KeyEqual = std::equal_to<Key>>
+class open_hash_set {
+public:
+    typedef Key key_type;
+    typedef Key value_type;
+    typedef size_t size_type;
+    typedef Hash hasher;
+    typedef KeyEqual key_equal;
+
+    class iterator {
+        value_type* ptr;
+    public:
+        iterator(value_type* ptr_) : ptr(ptr_) {}
+        value_type& operator*() const { return *ptr; }
+        value_type* operator->() const { return ptr; }
+        bool operator==(iterator x) const { return ptr == x.ptr; }
+        bool operator!=(iterator x) const { return ptr != x.ptr; }
+        bool operator>=(iterator x) const { return ptr >= x.ptr; }
+        bool operator<=(iterator x) const { return ptr <= x.ptr; }
+        bool operator>(iterator x) const { return ptr > x.ptr; }
+        bool operator<(iterator x) const { return ptr < x.ptr; }
+    };
+
+private:
+    hasher hash_instance;
+    key_equal equal_instance;
+    IsKeyNull null_instance;
```

Comment:
all these instances are going to occupy one byte of storage each, despite not having any state (this is because of the guarantee C++ give you that every object has a unique address). It may not matter, but one trick to make them not use any space is to derive from them, and rely on the empty base class optimization